design/bios-config.md

```diff
@@ -0,0 +1,151 @@
+
+## Motivation
+
+We need this document to specify the typical usage of BIOS YAML attributes in the BareMetalHost YAML. 
+
+
+### Goals
+
+1. To agree on the format of the attributes that are to be supported as a starting point
+2. To keep the attribute naming vendor agnostic
+
+### Non-Goals
+
+1. To list every attribute for every vendor exhaustively
+
+## Proposal
+The proposal is to set/unset the BIOS configuration values using vendor drivers available for each BMC type. For this purpose, new YAML attributes are to be introduced in the BareMetalHost spec. 
+ 
+### Implementation Details/Notes/Constraints 
+All the BIOS related attributes or fields will come under the sub-section called ```bios``` in spec. The values for each attribute can be a boolean or a string. The vendor driver type to implement this configuration is not to be specified separetely as it is already known from the bmc sub-section.
+Initially, the BIOS configuration can be applied using the controller types idrac and irmc
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BareMetalHost
+metadata:
+ name: bm0
+spec:
+ online: true
+ bmc:
+   address: idrac://192.168.122.1:6230/
+   credentialsName: bm0-bmc-secret
+ bootMACAddress: 52:54:00:b7:b2:6f	
+ bios:
+   EnableSriovGlobal: false
+   BootMode: uefi
+   EnableVirtualization: false
+   EnableHyperThreading: true
+   EnableAdjCacheLine: false
+```
+
+The user can only specify the BIOS values related to the BMC type being used as they will be validated accordingly.
+
+One gotcha to this is about hiding the vendor sprawl in the parameter names and coming up with generic titles for the config params. This would then need to be handled in the operator code (the vendor-specific implementation for Ironic API calls). Meaning, it will still be required to map these *generic* names to the actual parameter names in the API call for it to function.
+
+### Risks and Mitigations
+
+None
+
+## Design Details
+
+The BIOS config is only possible when the host is in cleaning state and the input is to be given as a set of cleaning steps.
+
+The code changes required would entail 
+- Creating a go struct with the BIOS configs in the ```ironic.go (pkg/provisioner/ironic/ironic.go)```
+
+- Creating a new function to validate config params and build a final JSON object, called ```buildBIOSConfig``` in ```ironic.go```
+
+- Creating a new function to build clean steps for BIOS config called ```buildBIOSCleanSteps``` in ```ironic.go```
+
+### Work Items
+
+- Extend the BareMetalHost CRD with the new parameters for iDRAC and Redfish BMC types.
+
+- Validation of input values in the YAML parameters
+
+- A function to handle the cleaning steps related to BIOS configuration and its implementation
+
+- Unit tests for all the work above
+
+
+### Dependencies
+
+- Ironic
+- Baremetal-operator
+
+### Test Plan
+
+- Unit tests for the functions
+
+- Mock deployment testing with vBMC
```

Comment:
Please don't take the vBMC path. We added support to cache settings in the redfish emulator, where as there is no ipmi compatible bios_interface in ironic.

Reply:
We weren't aware when we'll get the hardware. This was the plan to pursue in order to explore the dev env. We're hopefully getting the hardware soon, so we plan to use bare metal servers, ultimately, for testing. I'll make the changes.

Reply:
The redfish emulator (sushy-tools) can be used for basic testing still, just vBMC itself is problematic.

Reply:
@dtantsur Thank you for the additional clarity :)

Reply:
@dtantsur That is very helpful. Thank you!